cpp/include/cuvs/neighbors/ivf_pq.hpp

```diff
@@ -156,21 +144,79 @@ struct search_params : ann::search_params {
    * performance if tweaked incorrectly.
    */
   double preferred_shmem_carveout = 1.0;
-
-  /** Build a raft IVF_PQ search params from an existing cuvs IVF_PQ search params. */
-  operator raft::neighbors::ivf_pq::search_params() const
-  {
-    raft::neighbors::ivf_pq::search_params result = {
-      {}, n_probes, lut_dtype, internal_distance_dtype, preferred_shmem_carveout};
-    return result;
-  }
 };
 /**
  * @}
  */
 
+static_assert(std::is_aggregate_v<index_params>);
+static_assert(std::is_aggregate_v<search_params>);
+
+/** Size of the interleaved group. */
+constexpr static uint32_t kIndexGroupSize = 32;
+/** Stride of the interleaved group for vectorized loads. */
+constexpr static uint32_t kIndexGroupVecLen = 16;
+
+/**
+ * Default value returned by `search` when the `n_probes` is too small and top-k is too large.
+ * One may encounter it if the combined size of probed clusters is smaller than the requested
+ * number of results per query.
+ */
+template <typename IdxT>
+constexpr static IdxT kOutOfBoundsRecord = std::numeric_limits<IdxT>::max();
+
+template <typename SizeT, typename IdxT>
+struct list_spec {
+  using value_type = uint8_t;
+  using index_type = IdxT;
+  /** PQ-encoded data stored in the interleaved format:
+   *
+   *    [ ceildiv(list_size, kIndexGroupSize)
+   *    , ceildiv(pq_dim, (kIndexGroupVecLen * 8u) / pq_bits)
+   *    , kIndexGroupSize
+   *    , kIndexGroupVecLen
+   *    ].
+   */
+  using list_extents = raft::
+    extents<SizeT, raft::dynamic_extent, raft::dynamic_extent, kIndexGroupSize, kIndexGroupVecLen>;
+
+  SizeT align_max;
+  SizeT align_min;
+  uint32_t pq_bits;
+  uint32_t pq_dim;
+
+  constexpr list_spec(uint32_t pq_bits, uint32_t pq_dim, bool conservative_memory_allocation)
+    : pq_bits(pq_bits),
+      pq_dim(pq_dim),
+      align_min(kIndexGroupSize),
+      align_max(conservative_memory_allocation ? kIndexGroupSize : 1024)
+  {
+  }
+
+  // Allow casting between different size-types (for safer size and offset calculations)
+  template <typename OtherSizeT>
```

Comment:
No implementations should exist in `include/` anymore, just the prototype definitions for the functions / classes. ALL implementations should now be getting compiled in `src/`.